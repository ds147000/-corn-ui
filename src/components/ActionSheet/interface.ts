
export interface ShowActionSheetOption {
  /** 选项数组，Taro端数组长度最大为 6，H5端不限制 */
  list: string[]
  /** 选项文字颜色 */
  color: string
}

export type ShowActionSheet = (option: ShowActionSheetOption) => Promise<Taro.showActionSheet.SuccessCallbackResult>


export interface ActionSheetProps {
  /** 是否显示 */
  visible: boolean
  /** 标题 */
  title?: string
  /** 副标题 */
  subTitle?: string
  /** 标题水平方向 */
  titleAlign?: 'left' | 'center' | 'right'
  /** 是否显示头部 */
  showHead?: boolean
  /** 是否显示关闭按钮 */
  closable?: boolean
  /** 点击蒙层是否允许关闭 */
  maskClosable?: boolean
  /** 关闭回调 */
  onClose?(): void
}
