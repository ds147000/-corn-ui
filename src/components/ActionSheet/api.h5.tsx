import Protal from '../Portal'
import { Container } from './container.h5'

export interface SuccessCallbackResult {
  tapIndex: number
  errMsg: string
}

export interface ShowActionSheetOption {
  /** 选项数组，Taro端数组长度最大为 6，H5端不限制 */
  list: string[]
}

export type ShowActionSheet = (option: ShowActionSheetOption) => Promise<SuccessCallbackResult>



const showActionSheet: ShowActionSheet = async (option) => {
  // eslint-disable-next-line no-async-promise-executor
  return new Promise( async (res, rej) => {
    const cancel = (): void => rej({ tapIndex: -1, errMsg: '用户取消' })

    const success = (index: number): void => {
      res({
        /** 用户点击的按钮序号，从上到下的顺序，从0开始 */
        tapIndex: index,
        /** 调用结果 */
        errMsg: 'ok'
      })
    }

    try {
      await Protal.add(Container, { ...option, onSuccess: success, onCancel: cancel })

    } catch (error) {
      rej({ tapIndex: -1, errMsg: error.message })
    }
  })

}

export default showActionSheet
