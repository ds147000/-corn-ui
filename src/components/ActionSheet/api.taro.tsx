import Taro from '@tarojs/taro'
import { ShowActionSheet } from './interface'

const showActionSheet: ShowActionSheet = async (option) => {
  const _option: Taro.showActionSheet.Option = {
    itemList: option.list || [],
    itemColor: option.color
  }
  return Taro.showActionSheet(_option)
}

export default showActionSheet
