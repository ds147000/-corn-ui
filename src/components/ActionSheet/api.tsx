/* eslint-disable import/export */
// #if _APP === 'weapp'
import showActionSheet from './api.taro'

export default showActionSheet

// #else
// eslint-disable-next-line import/first
import showActionSheet from './api.h5'

export default showActionSheet
// #endif
