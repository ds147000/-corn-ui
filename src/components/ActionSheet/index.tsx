/* eslint-disable import/export */

import ActionSheet  from './actionSheet'

export { ActionSheetItem } from './item'
export default ActionSheet
