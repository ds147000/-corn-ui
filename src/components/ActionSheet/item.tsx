import React from 'react'
import { View } from '@tarojs/components'
import { ActionSheetItemProps } from './interface'

export const ActionSheetItem: React.FC<ActionSheetItemProps> = ({ text, align, onClick, suffix }) => {
  return (
    <View className="xrk-actionsheet-item" onClick={onClick}>
      <View className={`xrk-as-text xrk-actionsheet-${align}`}>
        {text}
      </View>
      {suffix}
    </View>
  )
}
