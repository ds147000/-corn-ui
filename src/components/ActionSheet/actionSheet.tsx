import { View } from '@tarojs/components'
import classNames from 'classnames'
import React, { useMemo } from 'react'
import Drawer from '../Drawer'
import Icon from '../Icon/index.h5'
import { ActionSheetItem } from './item'


export interface ActionSheetProps {
  /** 是否显示 */
  visible: boolean
  /** 标题 */
  title?: string
  /** 副标题 */
  subTitle?: string
  /** 标题水平方向 */
  titleAlign?: 'left' | 'center' | 'right'
  /** 是否显示头部 */
  showHead?: boolean
  /** 是否显示关闭按钮 */
  closable?: boolean
  /** 是否显示底部关闭按钮 */
  showCancel?: boolean
  /** 取消按钮的文案 */
  cancelText?: string
  /** 是否显示底部确定按钮 */
  showOk?: boolean
  /** 确定按钮的文案 */
  okText?: string
  /** 点击蒙层是否允许关闭 */
  maskClosable?: boolean
  /** 取消或关闭回调 */
  onClose?(): void
  /** 点击确认按钮回调 */
  onOk?(): void
  onHide?(): void
}

const ActionSheet: React.FC<ActionSheetProps> = ({
  title,
  subTitle,
  titleAlign = 'left',
  showHead = true,
  maskClosable = true,
  closable,
  onClose,
  showCancel,
  cancelText = '关闭',
  showOk,
  okText = '确定',
  onOk,
  children,
  ...props
}) => {
  const isShowFloor = showCancel || showOk
  const titleClass = useMemo(() => classNames(
    'xrk-actionsheet-header-center',
    `xrk-actionsheet-${titleAlign}`
  ), [ titleAlign ])


  return (
    <Drawer
      {...props}
      maskClosable={maskClosable}
      onClose={onClose}
      position="bottom" mask
    >
      <View className="xrk-actionsheet">
        {showHead && (
          <View className="xrk-actionsheet-header">
            <View className={titleClass}>
              {Boolean(title) && <View className="title">{title}</View>}
              {Boolean(subTitle) && <View className="sub-title">{subTitle}</View>}
            </View>
          </View>
        )}
        {closable && (
          <View
            className="xrk-actionsheet-close"
            data-testid="close"
            onClick={onClose}
          >
            <Icon name="delete" />
          </View>
        )}
        <View className="xrk-actionsheet-body">
          {children}
        </View>
        {isShowFloor && (
          <View className="xrk-actionsheet-floor">
            {showCancel && (<ActionSheetItem align={titleAlign} text={cancelText} onClick={onClose} />)}
            {showOk && (<ActionSheetItem align={titleAlign} text={okText} onClick={onOk} />)}
          </View>
        )}
      </View>
    </Drawer>
  )
}


export default ActionSheet
