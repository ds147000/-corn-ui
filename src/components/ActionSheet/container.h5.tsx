import React, { useState } from 'react'
import { ShowActionSheetOption } from './api.h5'
import ActionSheet from './actionSheet'
import { ActionSheetItem } from './item'

interface DomProps {
  onDestory(): void
  onCancel(): void
  onSuccess(index: number): void
}

export const Container: React.FC<DomProps & ShowActionSheetOption> = ({ onDestory, onSuccess, onCancel, list }) => {
  const [ show, setShow ] = useState(true)

  const _onClose = (): void => {
    setShow(false)
    onCancel()
  }

  const _onSuccess = (index: number): void => {
    setShow(false)
    onSuccess(index)
  }

  return (
    <ActionSheet
      showCancel
      showHead={false}
      visible={show}
      onClose={_onClose}
      onHide={onDestory}
      cancelText="取消"
    >
      {list.map((item, key) => (
        <ActionSheetItem align="center" key={item} text={item} onClick={(): void => _onSuccess(key)} />
      ))}
    </ActionSheet>
  )
}
