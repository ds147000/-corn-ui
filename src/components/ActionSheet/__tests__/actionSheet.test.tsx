/* eslint-disable no-magic-numbers */
import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import ActionSheet from '../actionsheet'

describe('ActionSheet', () => {
  test('显示', async () => {
    const screen = render(<ActionSheet visible />)

    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('隐藏', async () => {
    const screen = render(<ActionSheet visible={false} />)

    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('title', async () => {
    const screen = render(<ActionSheet visible title="大标题居中" />)
    screen.getByText('大标题居中')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('title of hide', async () => {
    const screen = render(<ActionSheet visible={false} title="大标题居中" />)

    await waitFor(() => expect(screen.queryByText('大标题居中')).toBe(null))
  })

  test('subTitle', async () => {
    const screen = render(<ActionSheet visible title="大标题居中" subTitle="辅助文字居中" />)
    screen.getByText('大标题居中')
    screen.getByText('辅助文字居中')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('subTitle of hide', async () => {
    const screen = render(<ActionSheet visible={false} title="大标题居中" subTitle="辅助文字居中" />)
    await waitFor(() => expect(screen.queryByText('大标题居中')).toBe(null))
    await waitFor(() => expect(screen.queryByText('辅助文字居中')).toBe(null))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('titleAlign: left', async () => {
    const screen = render(<ActionSheet visible title="大标题居左" titleAlign="left" />)
    screen.getByText('大标题居左')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('showHead: false', async () => {
    const screen = render(<ActionSheet visible title="大标题居中" subTitle="辅助文字居中" showHead={false} />)
    await waitFor(() => expect(screen.queryByText('大标题居中')).toBe(null))
    await waitFor(() => expect(screen.queryByText('辅助文字居中')).toBe(null))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('titleAlign: center', async () => {
    const screen = render(<ActionSheet visible title="大标题居中" titleAlign="center" />)
    screen.getByText('大标题居中')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('titleAlign: right', async () => {
    const screen = render(<ActionSheet visible title="大标题居右" titleAlign="right" />)
    screen.getByText('大标题居右')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('closable', async () => {
    const onClose = jest.fn()
    const screen = render(<ActionSheet visible closable title="大标题居右" onClose={onClose} />)
    screen.getByText('大标题居右')
    fireEvent.click(screen.getByTestId('close'))
    await waitFor(() => expect(onClose).toHaveBeenCalledTimes(1))
  })

  test('closable: false', async () => {
    const onClose = jest.fn()
    const screen = render(<ActionSheet visible title="大标题居右" onClose={onClose} />)
    const closeButton = screen.container.querySelector('xrk-actionsheet-close')
    expect(closeButton).toBe(null)
  })

  test('showButton', async () => {
    const screen = render(
      <ActionSheet
        visible
        title="我是大标题"
        subTitle="我是副标题"
        showCancel
        showOk
      />
    )
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('自定义文字', async () => {
    const screen = render(
      <ActionSheet
        visible
        title="我是大标题"
        subTitle="我是副标题"
        showCancel
        showOk
        cancelText="cancel"
        okText="ok"
      />
    )

    screen.getByText('cancel')
    screen.getByText('ok')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })
})
