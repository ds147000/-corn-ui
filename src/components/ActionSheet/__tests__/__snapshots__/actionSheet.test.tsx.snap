// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSheet showButton 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    >
      <div
        class="xrk-actionsheet-header"
      >
        <div
          class="xrk-actionsheet-header-center xrk-actionsheet-left"
        >
          <div
            class="title"
          >
            我是大标题
          </div>
          <div
            class="sub-title"
          >
            我是副标题
          </div>
        </div>
      </div>
      <div
        class="xrk-actionsheet-floor"
      >
        <div
          class="xrk-actionsheet-item"
        >
          <div
            class="xrk-as-text xrk-actionsheet-left"
          >
            关闭
          </div>
        </div>
        <div
          class="xrk-actionsheet-item"
        >
          <div
            class="xrk-as-text xrk-actionsheet-left"
          >
            确定
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionSheet showHead: false 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    />
  </div>
</div>
`;

exports[`ActionSheet subTitle 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    >
      <div
        class="xrk-actionsheet-header"
      >
        <div
          class="xrk-actionsheet-header-center xrk-actionsheet-left"
        >
          <div
            class="title"
          >
            大标题居中
          </div>
          <div
            class="sub-title"
          >
            辅助文字居中
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionSheet subTitle of hide 1`] = `<div />`;

exports[`ActionSheet title 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    >
      <div
        class="xrk-actionsheet-header"
      >
        <div
          class="xrk-actionsheet-header-center xrk-actionsheet-left"
        >
          <div
            class="title"
          >
            大标题居中
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionSheet titleAlign: center 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    >
      <div
        class="xrk-actionsheet-header"
      >
        <div
          class="xrk-actionsheet-header-center xrk-actionsheet-center"
        >
          <div
            class="title"
          >
            大标题居中
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionSheet titleAlign: left 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    >
      <div
        class="xrk-actionsheet-header"
      >
        <div
          class="xrk-actionsheet-header-center xrk-actionsheet-left"
        >
          <div
            class="title"
          >
            大标题居左
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionSheet titleAlign: right 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    >
      <div
        class="xrk-actionsheet-header"
      >
        <div
          class="xrk-actionsheet-header-center xrk-actionsheet-right"
        >
          <div
            class="title"
          >
            大标题居右
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionSheet 显示 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    >
      <div
        class="xrk-actionsheet-header"
      >
        <div
          class="xrk-actionsheet-header-center xrk-actionsheet-left"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionSheet 自定义文字 1`] = `
<div>
  <div
    class="xrk-drawer-mask xrk-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body xrk-drawer-bottom-show xrk-drawer-bottom"
    data-testid="body"
  >
    <div
      class="xrk-actionsheet"
    >
      <div
        class="xrk-actionsheet-header"
      >
        <div
          class="xrk-actionsheet-header-center xrk-actionsheet-left"
        >
          <div
            class="title"
          >
            我是大标题
          </div>
          <div
            class="sub-title"
          >
            我是副标题
          </div>
        </div>
      </div>
      <div
        class="xrk-actionsheet-floor"
      >
        <div
          class="xrk-actionsheet-item"
        >
          <div
            class="xrk-as-text xrk-actionsheet-left"
          >
            cancel
          </div>
        </div>
        <div
          class="xrk-actionsheet-item"
        >
          <div
            class="xrk-as-text xrk-actionsheet-left"
          >
            ok
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActionSheet 隐藏 1`] = `<div />`;
